design-proposals/instancetype.kubevirt.io/extend-preferSpread.md

```diff
@@ -0,0 +1,325 @@
+{
+  "cores": 2,
+  "model": "host-model",
+  "sockets": 2,
+  "threads": 1
+}
+```
+
+This design proposal aims to extend this implementation by introducing new configurables within a preference to control how vCPUs are spread with the goal of allowing vCPUs to also be spread to threads.
+
+## Motivation
+
+The current implementation of the `preferSpread` `preferredCPUTopology` is limited to spreading vCPUs across sockets and cores of the guest CPU topology. While a useful starting point it would be beneficial to also allow these vCPUs to be spread across all parts of the guest visible CPU topology including threads in order to mimic SMT within the guest.
+
+This is particularly useful for performance related workloads such as DPDK etc where we would like to make use of SMT on the host in order to map through sibling threads from pCPUs to guest thread siblings. Note that this use case requires `dedicatedCPUPlacement` and `guestMappingPassthrough` to be useful so the CPUManager can allocate complete pCPUs (via the `full-pcpus-only` option) and for these pCPU siblings to be mapped correctly as vCPU thread siblings to the guest.
+
+## Goals
+
+* Allow for vCPUs to be spread across all attributes of the guest visible CPU topology
+
+## User Stories
+
+* As an admin or VM owner I would like to provide preferences that spread vCPUs provided by an instance type over the guest visible sockets, cores and threads
+
+* As a VM owner I would like to consume preferences that that spread vCPUs provided by an instance type over the guest visible sockets, cores and threads
+
+## Repos
+
+* kubevirt/kubevirt
+
+* kubevirt/common-instancetypes
+
+# Design
+
+### `spec.cpu.spreadOptions`
+
+A new `spreadOptions` struct will be introduced to hold new configurables detailing how vCPUs should be spread over the guest visible CPU topology.
+
+```go
+type spreadAcross string
+
+type SpreadOptions struct {
+	// Across optionally defines how to spread vCPUs across the guest visible topology.
+	// Default: SocketsCores
+	//
+	//+optional
+	Across spreadAcross `json:"across,omitempty"`
+
+	// Ratio optionally defines the ratio to spread vCPUs across the guest visible topology.
+	// Default: 2
+	//
+	//+optional
+	Ratio uint32 `json:"ratio,omitempty"`
+}
+```
+
+The `ratio` option is a replacement for the original `spec.preferSpreadSocketToCoreRatio` configurable. The original being deprecated and discussed later in this proposal.
+
+A new `across` option will be introduced that will map to the following constants and behaviour.
+
+### `spreadAcross` `SocketsCores` (default)
+
+This option will spread vCPUs between sockets and cores. Given the current behaviour of `PreferSpread` this will remain the default.
+
+```go
+const SpreadAcrossSocketsCores SpreadAcross = "SocketsCores"
+```
+
+### `spreadAcross` `CoresThreads`
+
+This option will spread vCPUs between cores and threads with a max of 2 threads per core.
```

Comment:
This means ratio will be ignored?

Reply:
I think it's more that only a ratio of 2 is supported and will need to be enforced by the validation webhook.

I'll add a note.